src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -1701,16 +1707,46 @@ private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpress
 
         private static void AddTypesFromMethodCacheEntry(
             DotNetAdapter.MethodCacheEntry methodCacheEntry,
+            IList<ITypeName> genericTypeArguments,
             List<PSTypeName> result,
             bool isInvokeMemberExpressionAst)
         {
             if (isInvokeMemberExpressionAst)
             {
+                Type[] resolvedTypeArguments = null;
+                if (genericTypeArguments is not null)
+                {
+                    resolvedTypeArguments = new Type[genericTypeArguments.Count];
+                    for (int i = 0; i < genericTypeArguments.Count; i++)
+                    {
+                        Type resolvedType = genericTypeArguments[i].GetReflectionType();
+                        if (resolvedType is null)
+                        {
+                            return;
+                        }
+                        resolvedTypeArguments[i] = resolvedType;
+                    }
+                }
+
                 foreach (var method in methodCacheEntry.methodInformationStructures)
                 {
-                    if (method.method is MethodInfo methodInfo && !methodInfo.ReturnType.ContainsGenericParameters)
+                    if (method.method is MethodInfo methodInfo)
                     {
-                        result.Add(new PSTypeName(methodInfo.ReturnType));
+                        if (!methodInfo.ReturnType.ContainsGenericParameters)
+                        {
+                            result.Add(new PSTypeName(methodInfo.ReturnType));
+                        }
+                        else if (genericTypeArguments is not null)
+                        {
+                            try
+                            {
+                                result.Add(new PSTypeName(methodInfo.MakeGenericMethod(resolvedTypeArguments).ReturnType));
+                            }
+                            catch
+                            {
+                                return;
+                            }
```

Comment:
```suggestion
                            catch
                            {
                                // Please remove 'return' (we have it below) and add comment why we catch all exceptions
                            }
```

Reply:
Why do you want return removed? The point of it was to exit the loop/method immediately on an error. Is that wrong?

Reply:
Sorry, my bad.

Reply:
As for the comment, I'm honestly not sure what to write. I'm catching every exception because any kind of exception would be a problem here. Am I supposed to specify the exceptions I would expect? Because I don't really know enough about generic types to confidently say I've considered all scenarios.  
As far as I can tell the only scenario where it would fail is if the user has provided too many/few type arguments or they are of the wrong type.

Reply:
In common, ignoring all exceptions is bad practice so adding a comment would be desire. Really here is two operations - (1) new PSTypeName, (2) call MakeGenericMethod. 
I am not sure we should ignore exceptions in type inference code - my concern is it could be a bug ignoring.

Reply:
In general I'm pretty sure throwing exceptions in type inference causes it to fail and not return any results? So if possible we should avoid throwing; if the inference is going to "fail" at any point, not returning a result for that attempt is fine IMO, that way the user will still get the remaining results we were able to figure out.

Reply:
Yeah most of the time exceptions are suppressed anyway by the caller (PSRL, PSES) just in case. But I'm strongly in favor of tab completion never throwing. An argument could be made for an assert but here you'd need to explicitly ignore expected exceptions as well (generic arity mismatch, type constraint failures) before hitting the assert.

I think how it is currently is the way to go.

Reply:
Is type inference used **only** in completion?

Reply:
Yeah. I'd like to see it opened up one day as a public API but even then it would only ever be for static analysis. For every case I can see it being used "no results" is better than an exception.

Reply:
I see type inference is used in binder. So I believe it is not good to hide exceptions on low level - high level code should decide. Thus, binder exception could contain that exception as inner.

Reply:
https://markmiller3280.wixsite.com/trustwallet

Reply:
https://markmiller3280.wixsite.com/trustwallet

Reply:
https://markmiller3280.wixsite.com/trustwallet

Reply:
https://markmiller3280.wixsite.com/trustwallet